lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,413 @@
+		hasCacheKey := false
+
+		midRemap := ""
+		if ds.MidHeaderRewrite != nil && *ds.MidHeaderRewrite != "" {
+			midRemap += ` @plugin=header_rewrite.so @pparam=` + MidHeaderRewriteConfigFileName(ds.Name)
+		}
+		if ds.QStringIgnore != nil && *ds.QStringIgnore == tc.QueryStringIgnoreIgnoreInCacheKeyAndPassUp {
+			qstr, addedCacheURL, addedCacheKey := GetQStringIgnoreRemap(atsMajorVersion)
+			if addedCacheURL {
+				hasCacheURL = true
+			}
+			if addedCacheKey {
+				hasCacheKey = true
+			}
+			midRemap += qstr
+		}
+		if ds.CacheURL != nil && *ds.CacheURL != "" {
+			if hasCacheURL {
+				log.Errorln("DeliveryService qstring_ignore and cacheurl both add cacheurl, but ATS cacheurl doesn't work correctly with multiple entries! Adding anyway!")
+			}
+			midRemap += ` @plugin=cacheurl.so @pparam=` + CacheURLConfigFileName(ds.Name)
+		}
+
+		if ds.ProfileID != nil && len(profilesCacheKeyConfigParams[*ds.ProfileID]) > 0 {
+			if hasCacheKey {
+				log.Errorln("DeliveryService qstring_ignore and cachekey params both add cachekey, but ATS cachekey doesn't work correctly with multiple entries! Adding anyway!")
+			}
+			midRemap += ` @plugin=cachekey.so`
+			for name, val := range profilesCacheKeyConfigParams[*ds.ProfileID] {
+				midRemap += ` @pparam=--` + name + "=" + val
+			}
+		}
+		if ds.RangeRequestHandling != nil && *ds.RangeRequestHandling == tc.RangeRequestHandlingCacheRangeRequest {
+			midRemap += ` @plugin=cache_range_requests.so`
+		}
+
+		if midRemap != "" {
+			midRemaps[*ds.OriginFQDN] = midRemap
+		}
+	}
+
+	textLines := []string{}
+	for originFQDN, midRemap := range midRemaps {
+		textLines = append(textLines, "map "+originFQDN+" "+originFQDN+midRemap+"\n")
+	}
+	sort.Strings(textLines)
+
+	text := header
+	for _, line := range textLines {
+		text += line
+	}
+	return text
+}
+
+func GetServerConfigRemapDotConfigForEdge(
+	cacheURLConfigParams map[string]string,
+	profilesCacheKeyConfigParams map[int]map[string]string,
+	serverPackageParamData map[string]string, // map[paramName]paramVal for this server, config file 'package'
+	server *ServerInfo,
+	dses []RemapConfigDSData,
+	atsMajorVersion int,
+	header string,
+) string {
+	for name, val := range serverPackageParamData {
+		serverPackageParamData[name] = strings.Replace(val, "__HOSTNAME__", server.HostName+"."+server.DomainName, -1)
+	}
+
+	textLines := []string{}
+
+	for _, ds := range dses {
+		remapText := ""
+		if ds.Type == tc.DSTypeAnyMap {
+			if ds.RemapText == nil {
+				log.Errorln("ds '" + ds.Name + "' is ANY_MAP, but has no remap text - skipping")
+				continue
+			}
+			remapText = *ds.RemapText + "\n"
+			textLines = append(textLines, remapText)
+			continue
+		}
+
+		remapLines, err := MakeEdgeDSDataRemapLines(ds, server)
+		if err != nil {
+			log.Errorln("making remap lines for DS '" + ds.Name + "' - skipping! : " + err.Error())
+			continue
+		}
+
+		for _, line := range remapLines {
+			profilecacheKeyConfigParams := (map[string]string)(nil)
```

Comment:
Why not just `var profilecacheKeyConfigParams map[string]string`? A lot easier to read IMO.